pkg/services/authn/authnimpl/sync/oauth_token_sync.go

```diff
@@ -100,27 +103,49 @@ func (s *OAuthTokenSync) SyncOauthTokenHook(ctx context.Context, identity *authn
 		s.cache.Set(identity.ID, struct{}{}, getOAuthTokenCacheTTL(accessTokenExpires, idTokenExpires))
 		return nil
 	}
-	// FIXME: Consider using context.WithoutCancel instead of context.Background after Go 1.21 update
-	updateCtx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
-	defer cancel()
 
-	if err := s.service.TryTokenRefresh(updateCtx, token); err != nil {
-		if errors.Is(err, context.Canceled) {
-			return nil
-		}
-		if !errors.Is(err, oauthtoken.ErrNoRefreshTokenFound) {
-			s.log.Error("Failed to refresh OAuth access token", "id", identity.ID, "error", err)
-		}
+	lockKey := fmt.Sprintf("oauth-token-sync-%s", identity.ID)
+	_, err, _ = s.sf.Do(lockKey, func() (interface{}, error) {
+		s.log.Debug("Singleflight request for OAuth token sync", "key", lockKey)
 
-		if err := s.service.InvalidateOAuthTokens(ctx, token); err != nil {
-			s.log.Warn("Failed to invalidate OAuth tokens", "id", identity.ID, "error", err)
-		}
+		// FIXME: Consider using context.WithoutCancel instead of context.Background after Go 1.21 update
+		updateCtx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+		defer cancel()
+
+		if refreshErr := s.service.TryTokenRefresh(updateCtx, token); refreshErr != nil {
+			if errors.Is(refreshErr, context.Canceled) {
+				return nil, nil
+			}
+
+			token, _, err := s.service.HasOAuthEntry(ctx, identity)
+			if err != nil {
+				s.log.Error("Failed to get OAuth entry for verifying if token has already been refreshed", "id", identity.ID, "error", err)
+				return nil, err
+			}
 
-		if err := s.sessionService.RevokeToken(ctx, identity.SessionToken, false); err != nil {
-			s.log.Warn("Failed to revoke session token", "id", identity.ID, "tokenId", identity.SessionToken.Id, "error", err)
+			// if the access token has already been refreshed by another request (for example in HA scenario)
+			tokenExpires := token.OAuthExpiry.Round(0).Add(-oauthtoken.ExpiryDelta)
+			if !tokenExpires.Before(time.Now()) {
+				return nil, nil
+			}
```

Comment:
This seems like a repeat of what we are doing earlier in the function, but there we are checking id token also which I guess is not really needed here. But could still be a good idea to extract this to a function. wdyt?

Reply:
Yeah, maybe extracting the adjustment for the clock skew and expiry check make sense. I extracted those into the `getExpiryWithSkew` function.